httpbp/client_middlewares.go

```diff
@@ -0,0 +1,199 @@
+package httpbp
+
+import (
+	"io"
+	"net/http"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	retry "github.com/avast/retry-go"
+	"github.com/opentracing/opentracing-go"
+
+	"github.com/reddit/baseplate.go/retrybp"
+	"github.com/reddit/baseplate.go/tracing"
+	"github.com/sony/gobreaker"
+)
+
+// ClientMiddleware is used to build HTTP client middlewares by implementing
+// http.RoundTripper which http.Client accepts as Transport.
+type ClientMiddleware func(next http.RoundTripper) http.RoundTripper
+
+type roundTripper func(req *http.Request) (*http.Response, error)
+
+func (f roundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	return f(req)
+}
+
+// NewClient returns a standard HTTP client wrapped with the default
+// middlewares plus any additional client middlewares passed into this
+// function.
+func NewClient(config ClientConfig, middlewares ...ClientMiddleware) *http.Client {
```

Comment:
I am unsure of this formulation of a `NewClient` API.  This makes it unnecessarily tied to ClientConfig, for example.

The "business end" of this PR is the notion of middleware.  Let that be its legacy:

```
func WrapTransport(transport http.Transport, middleware ...ClientMiddleware) http.Transport {
  if base == nil {
    transport = http.DefaultTransport
  }
  for _, mw := range middleware {
    transport = mw(transport)
  }
  return transport
}

func (cc ClientConfig) Client() *http.Client {
  return &http.Client{
    Transport: WrapTransport(
      &http.Transport{
        MaxConnsPerHost: cc.MaxConnections,
      },
      ClientMonitor(cc.Slug),
      ClientRetry(),
      ClientCircuitBreaker(cc.CircuitBreaker),
    ),
  }
}
```

This makes the middleware usable even in contexts that don't have a ClientConfig, and it makes the ClientConfig responsible for knowing how to translate itself into appropriate middleware.

Reply:
That sounds good to me if you think we could have it in the reverse way where there is still a `NewClient(config Config)` but like you said have the middlewares not be coupled to the config struct but instead its primitives.

I like the idea to instantiate a `http.Client` from a config but I believe that is unlike any of our other APIs in baseplate.go